Optional certbot and support for manual certificates

sanyud24 opened the pull request:
issues:https:https://github.com/odk-x/tool-suite-X/issues/264
option to users to supply their own certificate

sanyud24 committed:
disable certbot

option to users to supply their own certificate

wbrunette commented:
@sanyud24 great start you will also need to create a new docker-compose-https.yml different name and have nginx look for the manual certificate.

Basically simply remove everything but the nginx section (and the things nginx depends on like volumes) and rename the new file. 

The launch stack launch at the end will need to use the new file instead so you will need another if statement there.

wbrunette commented:
@matt9j feel free to add any additional feedback.

wbrunette committed:
Merge pull request #29 from odk-x/development

release 2.1.8

wbrunette committed:
Merge pull request #30 from odk-x/demo

release 2.1.8

sanyud24 committed:
docker_compose_https_nginx

copied and changes the file name with nginx

matt9j commented on init-odkx-sync-endpoint.py:
```diff
@@ -83,15 +83,19 @@ def run_interactive_config():
-          --non-interactive".format(email, domain))
+        
+        manual_certificate = print("Do you wish to supply your own ssl certificate? [(Y)/n]")
+        if manual_certificate== "":
```
I don't believe this python code is not correct. Right now if the user enters n, no certificate will be generated. I may be misunderstanding, but I think this line should be something along the lines of `if proceed.strip().lower()[0] != "y":`, assuming you accept my suggestion that the default should be "n"

matt9j commented on init-odkx-sync-endpoint.py:
```diff
@@ -83,15 +83,19 @@ def run_interactive_config():
-          --keep-until-expiring \
-          --non-interactive".format(email, domain))
+        
+        manual_certificate = print("Do you wish to supply your own ssl certificate? [(Y)/n]")
```
I would suggest that the default behavior should be to automatically generate a certificate. Users having their own certificate is a special case. I'll defer to @wbrunette 's preference though.

sanyud24 replied:
Fixed now, default is N

matt9j commented on init-odkx-sync-endpoint.py:
```diff
@@ -83,15 +83,19 @@ def run_interactive_config():
+        if manual_certificate== "":
+            manual_certificate = "n"
+
+                os.system("sudo certbot certonly --standalone \
```
This entire function call should be one indent level less. In python the indentation is important and semantically part of the code.

sanyud24 replied:
yeah, missed it.. thanks for pointing it out

matt9j commented on init-odkx-sync-endpoint.py:
```diff
@@ -83,15 +83,19 @@ def run_interactive_config():
+        
+        manual_certificate = print("Do you wish to supply your own ssl certificate? [(Y)/n]")
+        if manual_certificate== "":
+            manual_certificate = "n"
```
Where is this assignment used?

matt9j commented on docker-compose-https_nginx.yml:
```diff
@@ -0,0 +1,18 @@
+      done & \
+      nginx -g \"daemon off;\" \
+      "
+    volumes:
```
How is the user-provided certificate added to nginx? Or is that not part of this pull request?

sanyud24 replied:
added now

matt9j reviewed (changes requested):
Hi @sanyud24, thanks for this contribution! I have a couple of questions opened in a review-- let me know if I can help clarify anything. Also more generally, were you able to test these changes? It is usually a good practice to test changes before submitting them, and add to your pull request how you did the testing. That can help me as a reviewer reproduce how things are supposed to work!

sanyud24 commented:
Really sorry to trouble you @matt9j  , I opened this  pullrequest to know if I'm on the right file , doing the right changes , I have my exams going on for a week now I'll correct everything in 2 -3 days after testing all the changes , I made some of the changes you suggested above a while ago but didn't commited them, I'll  commit after full testing .  Thankyou for the review.

matt9j commented:
> Really sorry to trouble you @matt9j , I opened this pullrequest to know if I'm on the right file , doing the right changes , I have my exams going on for a week now I'll correct everything in 2 -3 days after testing all the changes , I made some of the changes you suggested above a while ago but didn't commited them, I'll commit after full testing . Thankyou for the review.

No problem! Good luck with your exams, and I'm looking forward to seeing the finished product : ) You are indeed on the right track!

sanyud24 committed:
disable certbot

Inside the nginx_certificate.conf file user have to change the path of ssl certifcate and directory where ssl certificate private key is located .

sanyud24 commented:
I'm really not getting how to connect nginx code with dockerfile after storing private key and ssl certificate path . I could not test because of that , I'm sorry but I not getting anything to proceed with after this.

linl33 committed:
Merge pull request #32 from odk-x/development

move fixes for db-bootstrap to master

wbrunette committed:
Merge pull request #34 from odk-x/development

2.1.9 release

wbrunette committed:
Merge pull request #35 from odk-x/demo

2.1.9 release

sanyud24 committed:
Merge branch 'master' into disableCertbot

sanyud24 committed:
redid the whole script for manual certificates

fix typo

sanyud24 committed:
use correct nginx for certbot

sanyud24 commented:
@matt9j @wbrunette Fixed the changes as per the suggestions and with correct logic.
Please review the PR.

I did test the script with both manual script and using certbot, but since I didn't had an actual domain to point to, I couldn't test the nginx container on my own. But the configs and the volume mounting of custom certificates is tested and working.

wbrunette commented:
@matt9j for some reason it's not letting me request a re-review from you so consider this comment my request :)

matt9j commented on init-odkx-sync-endpoint.py:
```diff
@@ -48,65 +49,79 @@ def run_interactive_config():
         if enforce_https == "":
             enforce_https = "y"
-            enforce_https = enforce_https.lower().strip()[0]
+            enforce_https = enforce_https[0]
```
Why the index access here? I think it should be on line 52 instead? The intention I think was to allow someone to type "y" or "yes".

sanyud24 replied:
@matt9j yeah, the logic is kinda broken when the string is non-empty but not y or n, i.e 'yes' or 'no'. It'll keep retrying since the loop won't exit.
However, I didn't change the logic so it's out of scope of this PR. Would you still like me to fix it here or in another PR?

matt9j replied:
Leaving it here is okay.

matt9j commented on docker-compose-https.yml:
```diff
@@ -1,112 +1,12 @@
-  # A shared volume to hold automatically issued certs and share them
-  # between certbot and nginx.
-  https-certs-vol:
+      - ${CERT_FULLCHAIN_PATH}:/etc/nginx/certs/fullchain.pem
```
Do you know if these environment variables remain set if the system is rebooted and the swarm has to restart? I see they are set in the launch command in the python script, but I don't know if they'll be captured automatically by docker swarm.

sanyud24 replied:
Yeah, it should remain persisted since they are present in the generated compose config for the docker swarm.

matt9j reviewed (commented):
Generally this looks good to me. There is one issue with the environment variables you might want to double check. I don't have a test environment setup for this right now to test myself, but wanted to get some feedback to you without waiting another day :)

matt9j reviewed (approved):
Looks good to me per discussion in previous review.
